OCaml Cookbook

tmattio opened the pull request:
Very rough prototype to start discussions on the addition of a cookbook to the OCaml Learn area.

The goal is to provide short, task-oriented recipes to solve common programming problems, like reading from and writing to a file, querying an HTTP server, parsing JSON, etc.

The idea is that recipes can have variants, that can use different libraries. For instance, we can imagine variants of a recipe for parsing JSON that use Yojson, or Jsonm. Since the language documentation cannot use third-party libraries, the documentation currently doesn't give learners the right tools to get things done in OCaml, and solve real-world problems. The objective is to fill that gap.

tmattio committed:
Prototype OCaml Cookbook

sabine committed:
add a rule to crunch the cookbook directory

sabine commented:
Ideas for community contributions:
- [ ] previous / next links at the bottom would be nice
- [ ] filtering (categories / tags / package)
- [ ] would be nice to link to cookbook from a package's documentation page
- [ ] have a copy button to copy libraries / ppxes dune file
- [ ] add a table of contents for the categories and subcategories on `cookbook.eml`

sabine committed:
move position of cookbook

sabine committed:
side-by-side code view like Go by Example

sabine committed:
copy to clipboard, random number gen recipe

sabine committed:
hierarchy: category > task > recipe

- URL we serve this under is docs/cookbook/[task slug]/[recipe slug]
- multiple recipes per task
- recipe page displays links to other recipes for the same task

sabine committed:
bring back the styles on the input that I accidentally replaced by 'hidden'

sabine committed:
move cookbook to community section, refactor active_top_nav_item to be controlled by learn_layout.eml and community_layout.eml

tmattio commented:
Thanks a lot for taking over the cookbook, Sabine! I just ran it locally and it looks great, very exciting!!

Here are a few thoughts I had while looking at it:

- I would keep the cookbook in the Learn area. This is meant to be community driven, and so from a contributor perspective Community makes sense, but I think it's best to think of the sitemap from a visitor's point of view. In that case, I believe Learn is a better section for the cookbook.
- I would simplify the cookbook for now. For instance, I don't think we need to handle files (and certainly not multiple files), sections, explanations for code, etc. Instead, I would start with something deadsimple and only introduce things when we see that we need them (either because we received user feedback, or we have a design from Claire). 
- I think we can get rid of the cookbook categories yaml file. The title can be in the recipe themselves, and the categories (and their names) can be infered from the file structure.

If you agree with the second point, here's a suggestion for what a recipe could look like:

```
---
title: Generate random strings and arrays
packages:
- name: "mirage-crypto-rng"
  version: "0.11.2"
- name: "randomconv"
  version: "0.1.3"
---

\```ocaml
(* Initialize the RNG with the default entropy source. *)
let () = Mirage_crypto_rng_unix.initialize
          (module Mirage_crypto_rng.Fortuna)

(* Generate a `Cstruct.t` with random data. *)
let cstruct n = Mirage_crypto_rng.generate n

(* To get a random `char`, convert a random `int8`: *)
let char () = Char.chr (int8 ())

(* Use `Cstruct`'s conversion methods to obtain random byte arrays, bigarrays or strings. *)
let bytes n = cstruct n |> Cstruct.to_bytes
let bigarray n = cstruct n |> Cstruct.to_bigarray
let string n = cstruct n |> Cstruct.to_string

(* You can also create alphanumeric random characters: *)
let alphanum () =
  Char.chr (48 + Randomconv.int ~bound:74 cstruct)

(* Wrap your random value generator into a sequence to generate as many random values as you want. *)
let seq gen =
  Seq.unfold (fun () -> Some (gen (), ())) ()
let list n gen =
  seq gen |> Seq.take n |> List.of_seq
\```

```

sabine committed:
move to Learn area

F-Loyer committed:
Cookbook : encoding (#2097)

* Cookbook index : add encoding

* Cookbook encoding typos

* Add dune

* Update 00-camomile.md / remove tab

* Update 00-camomile.md / remove tab

* Update 00-camomile.md

* Apply suggestions from code review

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

---------

Co-authored-by: Frederic LOYER <frederic.loyer@club-internet.fr>
Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Cookbook : and Sorting list and arrays (#2098)

* Cookbook: add Sorting

* Update 00-stdlib.md / remove tabs

* Update 00-stdlib.md / remove tabs

---------

Co-authored-by: Frederic LOYER <frederic.loyer@club-internet.fr>

F-Loyer committed:
Cookbook textprocessing (#2104)

* Cookbook: text processing

* Cookbook: text processing (remove tabs)

* Cookbook: text processing (typo)

* Cookbook: text processing (typo)

* Cookbook: text processing (typo)

---------

Co-authored-by: Frederic LOYER <frederic.loyer@club-internet.fr>

F-Loyer committed:
Cookboot : Add a Database / ezsqlite entry (#2100)

* Cookbook : Database/ezsqlite

* Remove tabs

* Declare packages

* Indent error

* typo

* Update cookbook_categories.yml : fix folder name in cookbook_categories.yml

* Update cookbook_categories.yml - cookbook index fix

* Update 00-ezsqlite.md

---------

Co-authored-by: Frederic LOYER <frederic.loyer@club-internet.fr>

F-Loyer committed:
Cookbook concurrency : Lwt (#2107)

* Cookbook: add concurrency / Lwt

* Cookbook: add concurrency / Lwt - fix

* Cookbook: add concurrency / Lwt - fix

* Cookbook: add concurrency / Lwt - fix

---------

Co-authored-by: Frederic LOYER <frederic.loyer@club-internet.fr>

F-Loyer committed:
Cookbook caqti ppx rapper (#2108)

* Cookbook : Caqti and ppx_rapper

* Cookbook : Caqti and ppx_rapper - fix

* Cookbook : Caqti and ppx_rapper - fix

* Cookbook : Caqti and ppx_rapper - @ notation

* add query with WITH clause

* typo WITH/WHERE

* fix

* fix

* fix

* fix

* fix

---------

Co-authored-by: Frederic LOYER <frederic.loyer@club-internet.fr>

sabine committed:
simplify data format of cookbook files

sabine committed:
change data format to .ml file with YAML header, comments get displayed side-by-side with the code

F-Loyer committed:
Cookbook Ezsqlite - typo and rewriting (#2119)

Co-authored-by: Frédéric LOYER <frederic.loyer@club-internet.fr>

F-Loyer committed:
Cookbook Sorting - typo (#2118)

Co-authored-by: Frédéric LOYER <frederic.loyer@club-internet.fr>

sabine committed:
add dark mode styles

F-Loyer committed:
Cookbook : fix in Lwt (type mismatch with iter_s/iter_p functions) (#2127)

* Cookbook : fix in Lwt

* Update 00-lwt.ml

---------

Co-authored-by: Frédéric LOYER <frederic.loyer@club-internet.fr>

F-Loyer committed:
Update 00-caqti-ppx-rapper.ml - fix sqlite -> sqlite3 (#2126)

sabine committed:
rearrange category hierarchy and render all recipes for a task, name rercipes based on task and first listed package

sabine committed:
rebase

sabine committed:
add some tasks

sabine committed:
list the task separately in the table and all the recipes below the task

sabine committed:
add aria-label to make clear to screen-readers where the recipe link is going

Commit:
Added description field to tasks

Commit:
Task pages

Commit:
Task description

Commit:
Remove str and stdlib recipes

Commit:
Cookbook description text

Commit:
Formatting

sabine committed:
minor style changes

Commit:
Formatting

Commit:
Update CONTRIBUTING.md

Commit:
Fix tasks.yml processing

Commit:
Formatting

Commit:
Add Yaml deserializing recipe

Commit:
Refactor JSON tasks

Commit:
Text and formatting

Commit:
Review

Commit:
Remove sorting recipe

Commit:
Cleanup

Commit:
Delete postponed recipes

Commit:
Update localtime recipe

Commit:
Do not display package versions

Commit:
Formatting

Commit:
Display used libraries per package

Commit:
Display version as tested and category path

Commit:
Fix contribution text

Commit:
Fixes

Commit:
Formatting

Commit:
Fix category path

Commit:
Formatting

Commit:
Refactoring

Commit:
Add hl_yaml recipe

Commit:
Fix comments

Commit:
Formatting

Commit:
Formatting

Commit:
Review edits

Commit:
Fix typo

Commit:
Import FS tasks from PR 1400

Commit:
Bring back random generation

Commit:
Bring back random generation, con't

Commit:
Read text and gziped file

Commit:
Formatting

Commit:
Fix typo

Commit:
Formatting

Commit:
Remove some recipes

Commit:
Rebase

Commit:
Make task description optional

Commit:
Random values

Commit:
Formatting

christinerose committed:
line editing CONTRIBUTING.md
